### ISSUE TITLE
How should Element Timing interact with Shadow DOM?

### ISSUE THREAD

**Issue by npm1** [WICG/element-timing#3]
Element Timing exposes rendering timing information about images (and we want to add text soon). It has some attributes which contain useful information for attribution such as `element`. However, right now we are not specifying anything special about elements that are contained in shadow trees (not even if those are closed). What should we do in these cases? I guess we can split this question into two:
* What should be done for open trees?
* What should be done for closed trees?

Possible alternatives I can think of:
* Do no special handling: expose entries, with all the attributes set.
* Expose PerformanceElementTiming entries but with some attributes such as `element` set to null.
* Expose PerformanceElementTiming entries, but setting the `element` to the shadow host.
* Do not expose entries at all.
* Make it opt-in so that a developer must set a bit in order to get entries from shadow DOM.

I'd like some feedback here. Any thoughts? @annevk @rniwa
Also tagging @hayatoito who was the one who suggested the opt-in option for open shadow trees and said we should not expose the elements of closed shadow trees. I understand this constraint (and I will need to explicitly add checks to the spec) but we also want to enable developers to measure the performance of their website, and shadow trees can affect it, so ideally we can still expose valuable attribution.

**Comment by hayatoito** [comment]
@npm1 Thank you filing an issue!

@annevk @rniwa, could you help @npm1? We'd appreciate your thoughts to move Element Timing API forward.

cc: @chrishtr just in case.

**Comment by annevk** [comment]
It should be the same for open/closed and encapsulation should be preserved. So I think you should not expose anything from shadow trees, basically.

**Comment by npm1** [comment]
What is the benefit of 'encapsulation' when we are trying to measure the performance of a website? I think it would be very unfortunate if shadow trees cannot be measured and act like black boxes.

**Comment by annevk** [comment]
It's not about benefits, it's a design invariant. What you probably need to do is define some ElementInternals API to let a custom element decide what kind of timing it wants to expose to the outside.

cc @domenic @tkent-google

**Comment by npm1** [comment]
Ok, I think this is fine as long as the performance of slots  can be monitored in practice. In other words, following the example on [MDN](https://developer.mozilla.org/en-US/docs/Web/Web_Components/Using_templates_and_slots), if I have this slot in a shadow tree of a `my-paragraph` custom element:

```
<p><slot name="my-text">My default text</slot></p>
```

And I instantiate the element like this:

```
<my-paragraph>
	<span slot="my-text" id='theid'>Let's have some different text!</span>
</my-paragraph>
```

then `document.getElementById('theid')` works (i.e. returns the element), which means the element used to replace the slot is not part of the shadow DOM, correct?

In terms of how to spec it, the check would basically be to not observe `element` if "`element`'s` root is a shadow root"?

**Comment by annevk** [comment]
Correct / you might wanna check that the element's root is a document associated with a browsing context or some such instead. Depends a bit on your other preconditions.

**Comment by npm1** [comment]
Ok thanks @annevk! I'm not super familiar with Shadow DOM, could you elaborate on what's the design invariant for open shadow DOM here? Hayato suggested we could have an opt-in for open shadow, something like:

`observer.observe({entryTypes: ['element'], includesOpenShadow: true});`

Is that something you'd be strongly opposed to?

**Comment by annevk** [comment]
Yeah, we don't have that kind of thing for other APIs, such as mutation observers. If we did things like that it should be done consistently and also agreed upon by the wider Web Components community. Thus far we've been very conservative, including for open shadow roots.

**Comment by hayatoito** [comment]
@npm1 

> Ok, I think this is fine as long as the performance of slots can be monitored in practice.

I am afraid that that's not what we want here, in practice,

For example, given that:


`<my-paragraph>`'s shadow tree is:
```
  <p><slot name="my-text">My default text</slot></p>
```
`<my-app>`'s shadow tree is:
```
  <my-paragraph><span slot="my-text" id='theid'>foo1</span></my-paragraph>
  <my-paragraph><span slot="my-text" id='theid'>foo2</span></my-paragraph>
  <my-paragraph><span slot="my-text" id='theid'>foo2</span></my-paragraph>
  <p><slot name="my-text">My default text</slot></p>
```

Then, if a document tree contains only `<my-app></my-app>`:
```
  <my-app><p><slot name="my-text">bar</slot></p></my-app>
```

Then, we could get almost nothing useful information here; We only get info about "bar" text. We couldn't get any info about foo1, foo2, nor foo3. I'm afraid that this is not what we want.

**Comment by foolip** [comment]
@npm1 can you describe what is going to ship in Chromium, and can the spec be updated to reflect that if it's a conservative approach? I spotted a link to this issue in https://cs.chromium.org/chromium/src/third_party/blink/renderer/core/paint/image_element_timing.cc?l=113&rcl=85892919ac9a65a0a99924a43f7256c1f6bd36eb, and there is an actual check there that could be put in the spec somewhere to match.

**Comment by npm1** [comment]
We're ignoring shadow DOM elements even if they have their elementtiming attribute set, for now. I attempted to reflect this in the spec as follows:
In https://wicg.github.io/element-timing/#sec-process-loaded-image, step 4 checks whether the root is a document (for shadow DOM it would be a shadowRoot).

In https://wicg.github.io/element-timing/#sec-element-processing we iterate over doc's descendants (for shadow DOM I think it needs to be shadow-including descendant).

**Comment by toddreifsteck** [comment]
Although the current shape of the specification makes sense, I think there is a need to break down the use case for how a web site wants to measure elements created using shadow DOM and to ensure the patterns are well understood and can be used effectively.

@npm1 Has this work been done?

I recently spoke with the msn.com team and they stated they have been unable to use ElementTiming for some of their components due to this limitation are currently reliant on JavaScript performance.mark. Would be happy to put them in touch with folks to talk this through. Should it be a new issue given this issue seems clearly about ensuring the shadow DOM does not bleed over its encapsulation?

**Comment by yoavweiss** [comment]
@toddreifsteck - Yeah, it'd be great if you/they can open a new issue and outline the reasons as to why they can't use ElementTiming for their use case.

**Comment by nolanlawson** [comment]
Thinking about this problem (after discussion at TPAC), I see a few possible solutions:

-----

**Option 1: top-down**

As an analogue, consider [`getInnerHTML` from Declarative Shadow DOM](https://web.dev/declarative-shadow-dom/#serialization). This allows for:

```js
element.getInnerHTML({
  includeShadowRoots: true,
  closedRoots: [shadowRoot1, shadowRoot2, ...]
});
```

Patching this onto `PerformanceObserver`:

```js
const observer = new PerformanceObserver(/* ... */);
observer.observe({
  entryTypes: ["element"],
  includeShadowRoots: true,
  closedRoots: [shadowRoot1, shadowRoot2, ...]
});
```

This would recursively pierce into open shadow roots, but closed shadow roots would need to be explicitly enumerated.

**Pros**: Simplest to implement from the `PerformanceObserver` user's perspective, no need for component opt-in. Has precedent in `getInnerHTML`.
**Cons**: Open-shadow-root components cannot "opt out" – they are opted-in by default.

-----

**Option 2: bottom-up**

An alternative solution would be something like [cross-root aria delegation](https://github.com/leobalter/cross-root-aria-delegation), but allowing a component to opt-in to exposing its `elementtiming`s up the tree:

```js
customElements.define('x-foo', class extends HTMLElement {
  constructor() {
    super();
    this.attachShadow({ mode: "open", delegatesElementTiming: true })
      .innerHTML = '<div elementtiming="foo"></div>'
  }
});
```

**Pros:** Could piggy-back on existing cross-root aria delegation API (assuming it expands to allow non-aria attributes and to delegate multiple elements+attributes from one shadow root). Component authors opt-in.
**Cons**: Multiple nested shadow roots would need to delegate all the way up the tree, which is awkward to implement for both component authors and page authors.

-----

**Option 3: "it just works"**

A third option is to make `elementtiming` "just work," even within closed/open shadow roots.

**Pros:** Aligns with existing implementation of LCP. No extra work for web authors. In a sense, the `elementtiming` attribute is already an "opt-in," since there's no other reason for an element to use it except to get Element Timings.
**Cons:** No way for component authors to opt-out if (for whatever reason) they use the `elementtiming` attribute but don't actually want to get Element Timings. No other precedent (I can think of) for where an attribute on an element in a shadow root is "observable" by script with only access to the global scope.

-----

Based on this, my suggestion would be either option 3 or option 1.

**Comment by rniwa** [comment]
We definitely shouldn't do option (3). It fundamentally goes against the encapsulation principle of shadow DOM.

**Comment by annevk** [comment]
I'm also not a big fan of option 1 as written. We shouldn't have a separate API for open and closed. The design for declarative shadow trees doesn't have consensus.

**Comment by smaug----** [comment]
I think of these I prefer option 4, especially if we can figure out other use cases for the _expose_ attribute. It is very explicit.
The problem it does have is that if one uses _expose_ with a closed shadow root, it does break encapsulation.
But, in some sense _expose_ is like a C++ friend class, you let some specific other API to poke into the internals.

**Comment by nolanlawson** [comment]
@smaug---- In our research for cross-root ARIA [delegation](https://github.com/leobalter/cross-root-aria-delegation)/[reflection](https://github.com/Westbrook/cross-root-aria-reflection), we identified several attributes that would be useful to expose outside of shadow roots (`aria-*`, `role`, `for`, `list`, `popup`, etc.), but in each of those cases I think it would be best to only expose the attribute "one level up" from the shadow root. I admit I can't think of any attributes other than `elementtiming` where we really just want to expose it directly at the global level.

Although maybe there are some I missed? Or that may be defined in the future?

/cc @westbrook

**Comment by annevk** [comment]
What's a shadow element? A shadow host?

(A custom element taking care of exposing the encapsulated information from its shadow tree to its tree makes the most sense to me, which sounds like option 2 above.)

**Comment by yoavweiss** [comment]
> What's a shadow element? A shadow host?

Indeed, apologies.

**Comment by nolanlawson** [comment]
> Using the `elementtiming` attribute directly on a shadow element would expose that internal element's timing to the global timeline.

This would expose multiple elements from the shadow root, if there are multiple `elementtiming` attributes inside of the shadow root, correct?

```js
customElements.define('x-foo', class extends HTMLElement {
  constructor() {
    super();
    this.attachShadow({ mode: "open", delegatesElementTiming: true }).innerHTML = `
      <!-- foo and bar are both exposed -->
      <div elementtiming="foo"></div>
      <div elementtiming="bar"></div>
    `;
  }
});
```

(I'm assuming **option 2** above.)

**Comment by nolanlawson** [comment]
Ah, I misunderstood. I thought that (like cross-root ARIA), the timings would need to be re-exported up the tree (**option 2**). If they are exported globally (with namespacing), then that sounds more like **option 4**.

So could we expand option 4 with something like this?

```diff
customElements.define('x-foo', class extends HTMLElement {
  constructor() {
    super();
    this.attachShadow({
      mode: "open",
+      expose: {
+        attributes: ['elementtiming'],
+        namespace: 'someGloballyUniqueString'
+      }
    })
    .innerHTML = '<div elementtiming="foo"></div>'
  }
});
```

Would the `PerformanceElementTiming` have a `namespace` property then?

<details><summary>Context on Salesforce's use of shadow DOM</summary>

For context, the Salesforce CRM app is not a light DOM with a sprinkling of shadow roots; it's basically nothing but shadow roots (deeply nested). So for us to take advantage of element timing with **option 2**, we would need to re-export possibly dozens of times up the tree, which is not ideal. So my preference is probably **option 4**.

</details>

**Comment by yoavweiss** [comment]
The option you outlined above works for me. Are you interested in pushing its spec and/or implementation? I'm happy to help on either front! :)

**Comment by annevk** [comment]
I wonder what @rniwa thinks of option 4. Why is it better than 2? And how exactly would exposure happen? Would the data end up being copied somewhere or would there be some pointer into the shadow root available globally?

**Comment by rniwa** [comment]
Option 2 seems most straight forward & matches the other existing APIs.

Fundamentally, it's problematic when either users or components can't decide whether to export element timing or not. It seems particularly problematic for each component author to make this decision for all uses of that component regardless of where they appear.

**Comment by yoavweiss** [comment]
Thinking about this some more, I don't know that we even need an explicit opt-in, as the `elementtiming` attribute is already an opt-in in and of itself.

So, option 5:
* Any `elementtiming` attribute inside of a shadow tree would queue up an ElementTiming entry for the global observers. By default such entries would have an `element` attribute which points to the top-most shadow host*, and an `id` attribute which contains the shadow host's ID if any. The element timing's identifier would be properly reflected, and would provide developers with a clue as to which actual element queued up the entry.
  - "top most shadow host" - the ancestor shadow host that's part of the non-shadow DOM
* If the element itself provides an opt-in, the shadow element and its id would be exposed in the corresponding ElementTiming attributes.

That way we don't expose any details but the timing to the non-shadow DOM, unless the element explicitly opts-in for that,

This would match what's happening in practice for LCP, where the element's timing is reported, but without a pointer to the actual element. I guess we could use the same opt-in to enable exposing the actual element to LCP.

**Comment by yoavweiss** [comment]
> Option 2 seems most straight forward & matches the other existing APIs.

The cons of option 2 seems to outweigh their benefits, IMO.

> Fundamentally, it's problematic when either users or components can't decide whether to export element timing or not. It seems particularly problematic for each component author to make this decision for all uses of that component regardless of where they appear.

There is no point in adding `elementtiming` attributes if one does not want to expose those timings. What's the deveoper benefit of providing multiple, nested opt-ins?

**Comment by annevk** [comment]
A component could want to collect timing data locally and only give a more abstract view to the outside world. The whole point is that these are building blocks for elements, treating them as if they are normal elements just hidden from view defeats the point. And that's also why making `elementtiming` be some kind of magical feature is not a great idea. We want exposure to be a very conscious decision and we don't want to end up with a handful of differently named attributes that end up doing that I think.

**Comment by yoavweiss** [comment]
What do you mean by "locally"? Components don't have their own execution contexts, right?

**Comment by annevk** [comment]
Within the logic created for that component. And no, they don't.

**Comment by yoavweiss** [comment]
I don't believe it will be neither ergonomic nor desired for each component to start collecting its own metrics.
So it makes sense that if the component emits metrics, they'd be emitted to the global performance observers.

Under that assumption, we need to figure out if the `elementtiming` attribute is in itself a good enough opt-in, or we need developers to jump through more hoops (and if so, to what purpose).

**Comment by yoavweiss** [comment]
FYI - we're planning to discuss this on the [WG call today at 10am PT/7pm CET](https://lists.w3.org/Archives/Public/public-web-perf/2022Sep/0016.html)

**Comment by nolanlawson** [comment]
> The element timing's identifier would be properly reflected, and would provide developers with a clue as to which actual element queued up the entry.

I assume this means the value of the `elementtiming` attribute, not the ID of the element itself?

> If the shadow DOM provides an opt-in (e.g. when attached), the internal element and its id would be exposed

Is there value in exposing the ID if the element itself is exposed?

Overall Option 5 works from my POV. It is a bit un-ergonomic that one would have to deeply traverse all shadow roots and run `shadowRoot.querySelector("[elementtiming=foo]")` to find the `elementtiming=foo` element, but that does give an incentive for the opt-in. (Assuming open shadow – with closed shadow, the opt-in would be mandatory.)

**Comment by yoavweiss** [comment]
> I assume this means the value of the `elementtiming` attribute, not the ID of the element itself?

Indeed. 



> Is there value in exposing the ID if the element itself is exposed?

That's what we do for element timing in general, so doing the same seemed consistent. The value is in case the element itself was removed (and hence is null).

**Comment by jdapena** [comment]
Triage: discussion required for choosing one of the alternates.

**Comment by jasonwilliams** [comment]
> But this sort of creates a single global perspective that all observers of the page timings must agree to.

> For example, if a parent node opts to observe child timings, but then any child also opts to observe its own timings, right now the default proposal is to hide these timings from the parent container. This might be surprising if the two attributes aren't explicitly assigned by the same author.

So I guess this was meant to be analogous to something like Shadow DOM, whereby a container can decide to close off its implementation detail (in this case, internal paint events) from emitting outside of it’s boundary (regardless of whether a parent is interested or not).

> This might be surprising if the two attributes aren't explicitly assigned by the same author.

I think you’re right it can be surprising, it was set this way because a container by default was self-contained and didn’t leak out any detail unless explicitly opened (which is more likely when done by the same author).

I appreciate it does become more inaccurate the more you do this though as the outer container is discounting more and more content so giving false numbers. An example is a component wrapping ads which may need to be timed, the overall header (by a different author) only wants to time its own content.

## Responses

>But I would like that data to not be tied to specific elements or implementation details of my components; it would not be great if I change an internal detail of my WC only to have the timings change and it potentially considered a "breaking change" or something.


Yep this very much was in the spirit of “shadowed” mode for container timing, the idea that we can still measure overall paint time of a component, but overall we don’t care about the implementation detail and shouldn’t need to know *how* it’s put together. Also like you say, choosing elements within just causes breaking changes to happen and adds limits to the WC Author.


>Maybe we can address all concern together by changing from using attributes stored directly on the nodes, and instead just passing targetNode of interest to the PerformanceObserver -- similar to how MutationObserver observers work.

I think this will cause other problems that we need to account for, such as:

**What if the observer registers after the paint as happened?**
Currently the browser only keeps paint timing info for elements it knows are interesting and it achieves that using the attribute, there would need to be a new mechanism for doing that which won’t cause memory pressure (for buffered observers we need to keep these entries in memory until an interested observer starts).

**What if the observer registers before the paint has started?**
Whilst this is much less common It’s certainly possible for this to happen and we have codebases where the observer can start before JavaScript dynamically builds up the DOM with content. In this case there is no targetNode which exists yet.

We may need a way to register interest in elements or identifiers ahead of time out of tree but early enough for the browser to know when painting. This could be in the head or somewhere else. I don’t know if this is best discussed [over here](https://github.com/bloomberg/container-timing/issues/20) 

Good points from @justinfagnani  about performance.measure() scoping, I agree that having consistent semantics across web component performance APIs is ideal. But I think adding shadow-root scoping to marks/measures is a separate design space (probably a follow-up issue), same for the self-profiling JS API. For container timing, our immediate goal is to define how browser-collected paint data behaves with shadow DOM, so that developers get stable host-level timings.
